outputs/email_test.go

```diff
@@ -0,0 +1,166 @@
+}
+
+type emailRecorder struct {
+	addr string
+	auth smtp.Auth
+	from string
+	to   []string
+	msg  []byte
+}
+
+func TestEmailOutput_Send(t *testing.T) {
+	testCases := []struct {
+		name               string
+		lookupMXFunc       func(name string) ([]*net.MX, error)
+		emailConfig        *EmailOutput
+		expectedMessage    string
+		sendError          error
+		expectedError      error
+		expectedSentEmails int
+	}{
+		{
+			name: "happy path, with auth, server supports auth",
+			expectedMessage: fmt.Sprintf("To: anything@fubar.com\r\n" +
+				"From: sender@mailer.com\r\n" +
+				"Subject: email subject\r\n" +
+				"Content-Type: text/html; charset=UTF-8\r\n" +
+				"\r\n" +
+				"foo bar baz body\r\n"),
+			expectedSentEmails: 1,
+		},
+		{
+			name: "happy path, use multiple mx servers, no auth",
+			lookupMXFunc: func(name string) ([]*net.MX, error) {
+				return []*net.MX{
+					{
+						Host: "127.0.0.1",
+					},
+					{
+						Host: "128.0.0.1",
+					},
+				}, nil
+			},
+			expectedMessage: fmt.Sprintf("To: anything@fubar.com\r\n" +
+				"From: sender@mailer.com\r\n" +
+				"Subject: email subject\r\n" +
+				"Content-Type: text/html; charset=UTF-8\r\n" +
+				"\r\n" +
+				"foo bar baz body\r\n"),
+			expectedSentEmails: 1,
+		},
+		{
+			name:          "sad path, no recipients",
+			emailConfig:   &EmailOutput{Recipients: []string{}},
+			expectedError: errThereIsNoRecipient,
+		},
+		{
+			name:               "sad path, client uses AUTH, smtp server does not support AUTH",
+			sendError:          fmt.Errorf("smtp: server doesn't support AUTH"),
+			expectedError:      fmt.Errorf("smtp: server doesn't support AUTH"),
+			expectedMessage:    "",
+			expectedSentEmails: 0,
+		},
+		{
+			name: "sad path, use mx server, invalid recipient,",
+			emailConfig: &EmailOutput{
+				Name:       "my-email",
+				User:       "user",
+				Password:   "pass",
+				Host:       "127.0.0.1",
+				Port:       587,
+				Sender:     "sender@mailer.com",
+				Recipients: []string{"invalid recipient"},
+				UseMX:      true,
+			},
+			expectedSentEmails: 0,
+		},
+		{
+			name: "sad path, no mx server available",
+			lookupMXFunc: func(name string) ([]*net.MX, error) {
+				return []*net.MX{}, fmt.Errorf("no such host")
+			},
+			expectedSentEmails: 0,
+		},
+		{
+			name: "sad path, use mx servers, error sending email",
+			lookupMXFunc: func(name string) ([]*net.MX, error) {
+				return []*net.MX{
+					{
+						Host: "127.0.0.1",
+					},
+				}, nil
+			},
+			expectedMessage: fmt.Sprintf("To: anything@fubar.com\r\n" +
+				"From: sender@mailer.com\r\n" +
+				"Subject: email subject\r\n" +
+				"Content-Type: text/html; charset=UTF-8\r\n" +
+				"\r\n" +
+				"foo bar baz body\r\n"),
+			sendError:          fmt.Errorf("internal server error"),
+			expectedSentEmails: 0,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			var emailsSent int
+			f, r := mockSend(tc.sendError, &emailsSent)
+
+			var ec EmailOutput
```

Comment:
Please rename variable to reflect it's type

Reply:
fixed